### ISSUE TITLE
Building Indic Voices using Falcon

### ISSUE THREAD

**Issue by prajwaljpj** [festvox/festvox#24]
@saikrishnarallabandi I have been trying to build a synthesizer from these steps [](https://github.com/festvox/festvox/blob/master/voices/indic/hindi_male/run.sh)

the ./bin/do_build command is not able to find the option get_phseq. I have installed all my libraries using 
[fest_build.sh.txt](https://github.com/festvox/festvox/files/4970943/fest_build.sh.txt)

`
(falconenv) gnani@prajwal:~/projects/falcon/festvox/voices/indic/hindi_female/vox$ ./bin/do_build get_phseq 

./bin/do_build unknown option get_phseq
`

**Comment by saikrishnarallabandi** [comment]
Can you run ./bin/do_build label
instead of ./bin/do_build get_phseq

**Comment by prajwaljpj** [comment]
./bin/do_build label worked! Thanks!
However, this led to me another issue. I have been following [run.sh](https://github.com/festvox/festvox/blob/master/voices/indic/hindi_male/run.sh). at the point of training, 

```shell
(falconenv) gnani@prajwal:~/projects/falcon/festvox/voices/indic/hindi_female$ python local/train_phones.py --exp-dir exp/taco_one_phseq
Command line args:
 {'--checkpoint-dir': 'checkpoints',
 '--checkpoint-path': None,
 '--conf': None,
 '--exp-dir': 'exp/taco_one_phseq',
 '--gpu-id': '0',
 '--help': False,
 '--hparams': '',
 '--log-event-path': 'exp/log_tacotronOne'}
Using GPU  0
Traceback (most recent call last):
  File "local/train_phones.py", line 163, in <module>
    hparams.parse(args["--hparams"])
  File "/home/gnani/projects/falcon/festvox/src/falcon/hyperparameters.py", line 73, in __getattr__
    raise AttributeError("No such attribute: " + name)
AttributeError: No such attribute: parse
(falconenv) gnani@prajwal:~/projects/falcon/festvox/voices/indic/hindi_female$ 
```

**Comment by saikrishnarallabandi** [comment]
Can u comment dat line out

**Comment by prajwaljpj** [comment]
Worked!

Here,

```shell
(falconenv) gnani@prajwal:~/projects/falcon/festvox/voices/indic/hindi_female$ python local/train_phones.py --exp-dir exp/taco_one_phseq
Command line args:
 {'--checkpoint-dir': 'checkpoints',
 '--checkpoint-path': None,
 '--conf': None,
 '--exp-dir': 'exp/taco_one_phseq',
 '--gpu-id': '0',
 '--help': False,
 '--hparams': '',
 '--log-event-path': 'exp/log_tacotronOne'}
Using GPU  0
{'>': 0, '<': 1, 'pau': 2, 'e': 3, 'k': 4, 'ssil': 5, 'i': 6, 's': 7, 'A:': 8, 'nB': 9, 'tBh': 10, 'v': 11, 'E': 12, 'hv': 13, 'A': 14, 'p': 15, 'kh': 16, 'tB': 17, 'onas': 18, 'm': 19, 'enas': 20, '9r': 21, 'gh': 22, 'l': 23, 'aU': 24, 'tr': 25, 'i:': 26, 'dB': 27, 'u': 28, 'ch': 29, 'J': 30, 'j': 31, 'bh': 32, 'u:': 33, 'b': 34, 'g': 35, 'aI': 36, 'o': 37, 'tR': 38, 'c}': 39, 'A:nas': 40, 'i:nas': 41, 'rr': 42, 'dBh': 43, 'c': 44, 'ph': 45, 'N': 46, 'Jh': 47, 'dr': 48, 'dR': 49, 'aInas': 50, 'u:nas': 51, 'z': 52, 'rrh': 53, 'sr': 54, 'nr': 55, 'ow': 56, 'n~': 57, 'f': 58, '9r=': 59, 'x': 60, 'h': 61, 'ay': 62, 'q': 63, 'G': 64, 'unas': 65, 'ownas': 66, 'inas': 67, 'aUnas': 68, 'nX': 69, 'lr': 70}
Fnames file is  vox/fnames.train
Feat name is  phones
Fnames file is  vox/fnames.val
Feat name is  phones
0it [00:00, ?it/s]
Traceback (most recent call last):
  File "local/train_phones.py", line 258, in <module>
    clip_thresh=hparams.clip_thresh)
  File "local/train_phones.py", line 85, in train
    for step, (x, input_lengths, mel, y) in tqdm(enumerate(train_loader)):
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/tqdm/std.py", line 1130, in __iter__
    for obj in iterable:
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 345, in __next__
    data = self._next_data()
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 856, in _next_data
    return self._process_data(data)
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 881, in _process_data
    data.reraise()
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/_utils.py", line 395, in reraise
    raise self.exc_type(msg)
FileNotFoundError: Caught FileNotFoundError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/_utils/worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/gnani/.virtualenvs/falconenv/lib/python3.6/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/gnani/projects/falcon/festvox/src/falcon/utils/misc.py", line 278, in __getitem__
    return self.X[idx], self.Mel[idx], self.Y[idx]
  File "/home/gnani/projects/falcon/festvox/voices/indic/hindi_female/local/util.py", line 73, in __getitem__
    return populate_phonesarray(fname, self.feats_dir, self.feats_dict)
  File "/home/gnani/projects/falcon/festvox/voices/indic/hindi_female/local/util.py", line 18, in populate_phonesarray
    f = open(fname)
FileNotFoundError: [Errno 2] No such file or directory: 'vox/festival/falcon_phones/data_16714 00:00:03.59.feats'

```

Should it be 
```python
f = open(fname.split(" ")[0]+".feats")
```
Or is there something going wrong beforehand when fname is assigned?

**Comment by saikrishnarallabandi** [comment]
This is the issue:

vox/festival/falcon_phones/data_16714 00:00:03.59.feats

The filename is just the first column. May be I missed it while
copying from etc/fnamesNdurations


Can you do the following:

cp vox/fnames voc/fnames.copy

cat vox/fnames.copy | cut -d' ' -f 1 > vox/fnames

./vox/bin/traintest vox/fnames

cp vox/fnames.test vox/fnames.val

**Comment by prajwaljpj** [comment]
This worked! Thank you!

**Comment by saikrishnarallabandi** [comment]
Awesome. Just to confirm, Falcom voice worked where festvox traditional
voice failed?

**Comment by prajwaljpj** [comment]
No, Both worked. Traditional festvox did not work when i used [](http://festvox.org/bsv/c2176.html) . I think one of these commands in [](http://festvox.org/bsv/c2176.html) is the problem for hindi.  

```shell
$FESTVOXDIR/src/promptselect/make_nice_prompts find_freq TEXT0 TEXT1 TEXT2
$FESTVOXDIR/src/promptselect/make_nice_prompts make_freq_lex
$FESTVOXDIR/src/promptselect/make_nice_prompts find_nice TEXT0 TEXT1 ...
$FESTVOXDIR/src/promptselect/make_nice_prompts select_letter_n 

```

**Comment by saikrishnarallabandi** [comment]
Okay great

**Comment by prajwaljpj** [comment]
@saikrishnarallabandi Is there an inference script for Falcon? I have a couple of trained models. Was looking to test it and benchmark time taken on random news articles online. Is there a script where i could give a txt (or a file with txt) as an arg to the python script?

**Comment by saikrishnarallabandi** [comment]
Yes

The last line in this scrpt:
https://github.com/festvox/festvox/blob/master/voices/arctic/rms/run.sh

It uses phone sequence. If u r testing onthe test set created(fnames.test)
ntg needs to b done. If testing on a new text set, need to extract phones
from the text first. (Which can be done either mnually using lex.lookup or
using " .bin/do_build build_prompts " )

Lmk if this is unclear

**Comment by prajwaljpj** [comment]
I was trying your simple solution if testing with fnames.test
This is the command i used
```shell
python local/synthesize_phones.py exp/taco_one_phseq/checkpoints/checkpoint_step0400000.pth vox/ehmm/etc/txt.phseq.data exp/taco_one_phseq/tts_phseq
```

The results im getting is just a humm sound. 
Here are some samples and their alignments.
[falcinf2.zip](https://github.com/festvox/festvox/files/5050852/falcinf2.zip)

while inference i get this warning for almost every file
```shell
< pau e k ssil k i s A: nB ssil tBh A: ssil v E hv ssil A p nB e ssil kh e tB onas ssil m enas ssil k A: m ssil k A 9r ssil gh A 9r ssil l aU tr ssil 9r A hv A: ssil tBh A: pau > data_00001
Warning! doesn't seems to be converged
```

Also, data_00001 is not in my test set.  i do not have a txt.phseq.data.test.head
do i need to extract phones to run it on my fnames.text?

**Comment by prajwaljpj** [comment]
I pulled out the test set by 
```shell
while read fil ; do  grep $fil ehmm/etc/txt.phseq.data; done < fnames.test > ehmm/etc/txt.phseq.data.test
```
Still similar issue.

**Comment by saikrishnarallabandi** [comment]
Ah! What about the aligments during training?

they are put in exp/${exp_dir}/checkpoints/*alignment.png

${exp_dir} is the name of experiment directory you used while training the model

**Comment by prajwaljpj** [comment]
Prediction in training seems to be extremely good! 
here are some examples with alignments
[falc_train_sample.zip](https://github.com/festvox/festvox/files/5051349/falc_train_sample.zip)

**Comment by saikrishnarallabandi** [comment]
ok. there is a known issue that leads to such outputs when the decoder is in eval mode

https://github.com/r9y9/tacotron_pytorch/issues/20

I followed this and just put encoder, postnet in eval mode but left decoder as is. Can you confirm that this is the case in the script you are using for synthesis?

**Comment by saikrishnarallabandi** [comment]
https://github.com/r9y9/tacotron_pytorch/issues/11 

this too.

One difference from the original paper is that in the original paper only the last frame is used to predict the next r frames. In Ryuichi's implementaiton, he uses all r frames and this might be an issue. I follow that same implementation. I also have an alternative implementation that uses only last frame which can be found [here](https://github.com/festvox/festvox/blob/master/voices/arctic/rms/local/train_phones_finalframe.py)

**Comment by prajwaljpj** [comment]
Very interesting issue!
It worked with your solution! Thank you! 
It seems to have a lot of trailing silence though.
Here are some samples
[falcon_inference_working.zip](https://github.com/festvox/festvox/files/5052469/falcon_inference_working.zip)

**Comment by saikrishnarallabandi** [comment]
Awesome. Great.

Ya may be silences are due to model not predicting 'End Of Sentences'
correctly.

Original paper talks that the training procedure should alleviate this. I
personally found that this goes away when I train longer. But you seem to
have already trained for 400K timestepops

Anyway, cool so Falcon worked. Good to know

**Comment by prajwaljpj** [comment]
Thank you very much for your elaborate help! I will take a look at the paper! 
I might need yout kind help in [this](https://github.com/festvox/festvox/issues/24#issuecomment-670372512) for festvox model i trained. Maybe its more relevant on the festvox/festival#13
Im still a beginner in Scheme programming and not sure how to implement lex.lookup through festival.

**Comment by prajwaljpj** [comment]
> [r9y9/tacotron_pytorch#11](https://github.com/r9y9/tacotron_pytorch/issues/11)
> 
> this too.
> 
> One difference from the original paper is that in the original paper only the last frame is used to predict the next r frames. In Ryuichi's implementaiton, he uses all r frames and this might be an issue. I follow that same implementation. I also have an alternative implementation that uses only last frame which can be found [here](https://github.com/festvox/festvox/blob/master/voices/arctic/rms/local/train_phones_finalframe.py)

Does using only the last frame result in better alignments? Im not getting good alignments on most of my test set. Here are some examples
[alignments.zip](https://github.com/festvox/festvox/files/5056915/alignments.zip)
Can this be averted by using a larger corpus? I have a lot of Hindi text data but around 45 hours of clean Hindi voice data

**Comment by saikrishnarallabandi** [comment]
Usually larger training data helps, yes

**Comment by prajwaljpj** [comment]
I am running into a strange issue where running inference over a sentence say,

( sampletwo_001 "उन्होंने कहा कि पहले बिहार पुलिस का एफआइआर मुंबई पुलिस के पास ट्रांसफर हो इसके बाद यदि महाराष्ट्र सरकार सीबीआई जांच की मंजूरी दे तो तभी सीबीआई जांच हो" )

does not work whereas running pieces of the same sentence work.

( piece_004 "उन्होंने कहा" )
( piece_005 "पुलिस का" )

Also tried jumbling up the words which works too.

( piece_000 "मंजूरी बिहार उन्होंने हो कहा कि पहले पुलिस का तो तभी" )

I have attached the necessary files and their alignments below.
[samples.zip](https://github.com/festvox/festvox/files/5073167/samples.zip)

I was hoping if you have any useful insights over this issue.

**Comment by saikrishnarallabandi** [comment]
So I have tried generating that sentence using three voices I built before.
1 male and 2 female

The samples:
http://tts.speech.cs.cmu.edu/rsk/misc_stuff/tts_falcondebudding/sampletwo_001.wav
http://tts.speech.cs.cmu.edu/rsk/misc_stuff/tts_falcondebudding/synth_miscvoice2/sampletwo_001.wav
<http://tts.speech.cs.cmu.edu/rsk/misc_stuff/tts_falcondebudding/synth_miscvoice3/sampletwo_001.wav>
http://tts.speech.cs.cmu.edu/rsk/misc_stuff/tts_falcondebudding/synth_miscvoice3/sampletwo_001.wav

Since these seem to work, I think something with respect to training the
model longer is the issue. But I remember seeing you train the model till
400K timesteps.
Is this dataset you are using open? If I can look into the data and build
model myself that might be very helpful to debug this issue

**Comment by prajwaljpj** [comment]
Interesting! I tried synthesizing with model of timestep 400k, 300k, 100k, and 50k.
All result in a similar issue. But at 100k it was able to predic End of Sentence for some of the samples. 
This is the script i use for inference
```shell
HEAPSIZE=20000000
VOXDIR=vox
EHMMDIR=$FESTVOXDIR/src/ehmm
PROMPTFILE=$1
cd vox
$ESTDIR/../festival/bin/festival --heap $HEAPSIZE -b festvox/build_clunits.scm '(build_prompts "'$PROMPTFILE'")'
$ESTDIR/../festival/bin/festival -b $EHMMDIR/bin/phseq.scm '(phseq "'$PROMPTFILE'" "ehmm/etc/sample.phseq.data")'
cd ..
python local/synthesize_phones.py exp/taco_one_phseq/checkpoints/checkpoint_step0450000.pth $VOXDIR/ehmm/etc/sample.phseq.data exp/taco_one_phseq/sample_out
```
Where i feed sample.done.data which has

```lisp
( sampletwo_001 "उन्होंने कहा कि पहले बिहार पुलिस का एफआइआर मुंबई पुलिस के पास ट्रांसफर हो इसके बाद यदि महाराष्ट्र सरकार सीबीआई जांच की मंजूरी दे तो तभी सीबीआई जांच हो" )
( piece_000 "मंजूरी बिहार उन्होंने हो कहा कि पहले पुलिस का तो तभी" )
( piece_004 "उन्होंने कहा" )
( piece_005 "पुलिस का" )
```
Am i missing something here? Looks like im missing something in my inference script.

Also, sorry, unfortunately I cannot share the data because it is proprietary.

**Comment by prajwaljpj** [comment]
Some more information,
I split the sentences to words and ran inference on each word one by one.
```lisp
( piece_01 "उन्होंने" )
( piece_02 "कहा" )
( piece_03 "कि" )
( piece_04 "पहले" )
( piece_05 "बिहार" )
( piece_06 "पुलिस" )
( piece_07 "का" )
( piece_08 "एफआइआर" )
( piece_09 "मुंबई" )
( piece_10 "पुलिस" )
( piece_11 "के" )
( piece_12 "पास" )
( piece_13 "ट्रांसफर" )
( piece_14 "हो" )
( piece_15 "इसके" )
( piece_16 "बाद" )
( piece_17 "यदि" )
( piece_18 "महाराष्ट्र" )
( piece_19 "सरकार" )
( piece_20 "सीबीआई" )
( piece_21 "जांच" )
( piece_22 "की" )
( piece_23 "मंजूरी" )
( piece_24 "दे" )
( piece_25 "तो" )
( piece_26 "तभी" )
( piece_27 "सीबीआई" )
( piece_28 "जांच" )
( piece_29 "हो" )
```
I have attached the results below
[words.zip](https://github.com/festvox/festvox/files/5082656/words.zip)

It is able to synthesize most words.

**Comment by prajwaljpj** [comment]
Also, I tried a random sentence from a Hindi news article
```lisp
( sentence_004 "बेलारूस के दीर्घकालिक राष्ट्रपति लुकाशेंको ने पोलैंड और लीथूआनिया में चल रहे नेटो के सैन्य अभ्यास पर चिंता ज़ाहिर करते हुए पश्चिमी देशों के सैन्य गठबंधन की आलोचना भी की है" )
```
This is the phoneme sequence
```shell
< pau b e l A: 9r u: s ssil k e ssil dB i: 9r gh A k A: l i k ssil 9r A: sr tr 9r A p tB i ssil l u k A: c} e N k o ssil nB e ssil p o l aI nr dr ssil aU 9r ssil l i: tBh u: A: nB i j A: ssil m enas ssil c A l ssil 9r A hv e ssil nB e tr o ssil k e ssil s aI nB j ssil A bh j A: s ssil p A 9r ssil c i nB tB A: ssil z A: hv i 9r ssil k A 9r tB e ssil hv u e ssil p A c} c i m i: ssil dB e c} onas ssil k e ssil s aI nB j ssil g A tR b A nB dBh A nB ssil k i: ssil A: l o c nB A: ssil bh i: ssil k i: ssil hv aI pau > (sentence_004).
```
The synthesis is failing after the word नेटो (नेटो inclusive)

Now, when I synthesize just the word नेटो
```lisp
( senp_01 "नेटो" )
```
It is able to synthesize correctly.
[neto.zip](https://github.com/festvox/festvox/files/5083504/neto.zip)

I have attached the necessary files for your reference. Any thoughts?

**Comment by saikrishnarallabandi** [comment]
If the training data has only short sequences, and test has longer ones it
is not an easy task for a Seq2Seq model. This is not a problem for
Clustergen since its not a seq2seq. But tacotron might get affected

**Comment by prajwaljpj** [comment]
The training data has long sequences (from 2 to 25 words). But the audio duration seems to be under 15 secs.

**Comment by prajwaljpj** [comment]
@saikrishnarallabandi How do I fine-tune a model with new dataset? 
I already have a trained model. Found more data. What s the fastest way to fine-tune my existing model?

**Comment by prajwaljpj** [comment]
Sorry, i did not have a free gpu to do this test till now. This worked! Thank you. I am able to continue training. I have also noticed that my dataset consists of a bit of leading silence and a lot of trailing silences. Surprisingly 
```shell
./bin/prune_silence wav/*.wav
./bin/prune_middle_silence wav/*.wav
```
did not have any effect on it and i didnt care to check whether it actually removed silences. I have seen that tacotron cannot generate attentions when the data contains silences. Will be training a model by manually removing these silences and will keep you updated. Thank you again!
Closing issue
